File: Algorithm/Complexity/python/binary.py
import random 
from sys import stdin

def readLine():
	return stdin.readline().strip()

def readInt():
	return int(readLine())

def binary(array, start, end, objective):
    print(f'Finding {objective} between {array[start]} and {array[end - 1]}')
    if start > end:
        return False
    
    middle = (start + end)//2
    
    if array[middle] == objective:
        return True
    elif array[middle] < objective:
        return binary(array, middle + 1, end, objective)
    else: 
        return binary(array, start, middle - 1, objective)

if __name__ == '__main__':
    print("size:")
    size_array = readInt()
    print("objective:")
    objective = readInt()
    
    array = sorted([random.randint(0, 100) for i in range(size_array)])

    number = binary(array, 0, len(array), objective)
    print(array)
    print(f'Element {objective} {"IF it's here" if number else "It's NOT here"} in the list')

File: Algorithm/Complexity/cpp/binary.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cstdlib>
#include <ctime>
#include <limits>

using namespace std;

bool binary(const vector<int>& arr, int start, int end, int objective) {
    if (start > end) return false;

    int middle = (start + end) / 2;
    cout << "Finding " << objective 
        << " between " << arr[start] 
        << " and " << arr[end - 1] << endl;

    if (arr[middle] == objective) {
        return true;
    } else if (arr[middle] < objective) {
        return binary(arr, middle + 1, end, objective);
    } else {
        return binary(arr, start, middle - 1, objective);
    }
};

int main() {
    int size, objective;
    cout << "Size: ";
    cin >> size;
    // cin.ignore(numeric_limits<streamsize>::max(), '\n');
    cout << "Objective: ";
    cin >> objective;

    srand(time(0));
    vector<int> arr(size);
    for (int i = 0; i < size; i++){
        arr[i] = rand() % 101;
    }

    sort(arr.begin(), arr.end());

    bool found = binary(arr, 0, size - 1, objective);

    cout << "Array: ";
    for (int num : arr ) cout << num << " ";
    cout << endl;

    cout << "Element " << objective
        << (found ? " Is Here" : "Is NOT here")
        << " in the list" << endl; 

    return 0;
}